contracts/interchain-token-service/tests/flow_limit.rs

```diff
@@ -0,0 +1,290 @@
+            destination_address: recipient,
+            amount,
+            data: None,
+        }),
+    };
+    let payload = msg.abi_encode(env).unwrap();
+    let payload_hash: BytesN<32> = env.crypto().keccak256(&payload).into();
+
+    let message_id = Address::generate(env).to_string();
+
+    let messages = vec![
+        &env,
+        GatewayMessage {
+            source_chain: source_chain.clone(),
+            message_id: message_id.clone(),
+            source_address: source_address.clone(),
+            contract_address: client.address.clone(),
+            payload_hash,
+        },
+    ];
+
+    (source_chain, message_id, source_address, payload, messages)
+}
+
+#[test]
+fn set_flow_limit_succeeds() {
+    let (env, client, _, _, _) = setup_env();
+    let token_id = BytesN::from_array(&env, &[1; 32]);
+
+    assert_eq!(client.flow_limit(&token_id), None);
+
+    assert_invoke_auth_ok!(
+        client.operator(),
+        client.try_set_flow_limit(&token_id, &TEST_FLOW_LIMIT)
+    );
+
+    assert_eq!(client.flow_limit(&token_id), TEST_FLOW_LIMIT);
+
+    goldie::assert!(events::fmt_last_emitted_event::<FlowLimitSetEvent>(&env));
+}
+
+#[test]
+#[should_panic(expected = "Error(Contract, #20)")] // FlowLimitExceeded
+fn zero_flow_limit_freezes_token() {
+    let (env, client, gateway_client, _, signers) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, _) = setup_flow_limit(&env, &client);
+
+    client.mock_all_auths().set_flow_limit(&token_id, &Some(0));
+
+    let amount = 1;
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, amount);
+    approve_gateway_messages(&env, &gateway_client, signers, messages);
+
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+}
+
+#[test]
+fn set_flow_limit_fails_invalid_amount() {
+    let (env, client, _, _, _) = setup_env();
+    let token_id = BytesN::from_array(&env, &[1; 32]);
+
+    let invalid_limit = Some(-1);
+
+    assert_contract_err!(
+        client
+            .mock_all_auths()
+            .try_set_flow_limit(&token_id, &invalid_limit),
+        ContractError::InvalidFlowLimit
+    );
+}
+
+#[test]
+fn flow_limit_resets_after_epoch() {
+    let (env, client, gateway_client, _, signers) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, _) = setup_flow_limit(&env, &client);
+
+    let amount = TEST_FLOW_LIMIT.unwrap();
+
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, amount);
+    approve_gateway_messages(&env, &gateway_client, signers.clone(), messages);
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+    assert_eq!(client.flow_in_amount(&token_id), amount);
+
+    let current_timestamp = env.ledger().timestamp();
+    env.ledger().set_timestamp(current_timestamp + EPOCH_TIME);
```

Comment:
```suggestion
    env.ledger().set_timestamp(env.ledger().timestamp() + EPOCH_TIME);
```
also fine so it's a bit concise